magefiles/rulesengine/readme.md

```diff
@@ -0,0 +1,211 @@
+ a Rule Catalog. 
+
+ The logic in this rule is straightforward: WHEN the RepoName is infra-deployments, THEN execute ginkgo with the specific label filter.
+
+ ```go
+ // Example rule for infra-deployments using anonymous functions embedded in the rule
+var InfraDeploymentsTestRulesCatalog = rulesengine.RuleCatalog{
+	rulesengine.Rule{Name: "Infra Deployments Default Test Execution",
+		Description: "Run the default test suites which include the demo and components suites.",
+		Condtion:    rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {
+			if rctx.RepoName == "infra-deployments" {
+				return true
+			}
+			return false
+		}),
+		Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {
+			rctx.LabelFilter = "e2e-demo,rhtap-demo,spi-suite,remote-secret,integration-service,ec,build-templates,multi-platform"
+			return ExecuteTestAction(rctx)
+		})}},
+
+} 
+ 
+ ```
+
+ ### Registering a rule catalog to the engine
+
+ Here I've created a catagory called tests within the MageEngine and I've assigned infra-deployments catalog to the infra-deployments key
+ under the tests category
+
+ ```go
+ var MageEngine = rulesengine.RuleEngine{
+	"tests": {
+		"infra-deployments": testselection.InfraDeploymentsTestRulesCatalog,
+	},
+}
+ ```
+
+ ### Executing the MageEngine within a mage file
+
+ This will cause the engine to load and iterate over the rules of the category test and apply any that had their conditions match.
+
+ ```go
+
+    rctx := rulesengine.NewRuleCtx()
+	rctx.DryRun = true
+
+	err = engine.MageEngine.RunRulesOfCategory("tests", rctx)
+ ```
+
+ ### Creating a complex rule using a mixture of conditional filters
+
+ In this example, the release-catalog-repo has a slightly more complex set of rules where depending on certain conditions
+ met by the CI job the test should execute with a different set of ginkgo label filters. So you will see a mixture of the 
+ Any/All/None conditional filters
+
+ The first rule's logic is: WHEN EITHER RepoName equals release-service-catalog AND NOT pr paired OR RepoName equals release-service-catalog
+ AND is a rehearse job AND NOT pr paired THEN run ginkgo with release-pipelines label filter
+
+ The second rule's logic is: WHEN RepoName equals release-service-catalog AND pr is paired AND NOT a rehearse job THEN run ginkgo with 
+ release-pipelines && !fbc-tests label filters
+
+
+ ```go
+
+ rulesengine.Rule{Name: "Release Catalog Test Execution",
+		Description: "Runs release catalog tests on release-service-catalog repo on pull/rehearsal jobs.",
+		Condtion:    rulesengine.Any{rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.None{rulesengine.ConditionFunc(isPaired)}}, rulesengine.All{rulesengine.ConditionFunc(releaseCatalogRepoCondition), rulesengine.ConditionFunc(isRehearse)}, rulesengine.None{rulesengine.ConditionFunc(isPaired)}},
```

Comment:
From the logic before - if rulesengine.ConditionFunc(isRehearse) is true, no test need to be run. Need to move "rulesengine.ConditionFunc(isRehearse)" to rulesengine.None{..}.

Reply:
Yes, I think you're right in the case of putting the settings in place because the setting up of specific values is a little more complex. I was thinking we could create those as separate rules. These example rules I was focused more on on test execution so when your team uses `release-pipelines` vs `release-pipeliens && !fbc-tests` label filter

That said, I reviewed the logic in the example and in the actual mage code and I think I overthought the conditional logic to try to catch for corner cases that may never happen in CI execution. I simplified the logic and it should follow what we have in the mage file. Let me know if it is clearer